http/cookies.lua

```diff
@@ -0,0 +1,356 @@
+	local n = #s_cookies
+	for index, value in pairs(s_cookies) do
+		if cookie_hashes[value] then
+			s_cookies[index] = nil
+			local by_domain = s_cookies[value.domain]
+			local by_path = by_domain[value.path]
+			by_path[value.key] = nil
+			if not next(by_path) then
+				by_domain[value.path] = nil
+				if not next(by_domain) then
+					s_cookies[value.domain] = nil
+				end
+			end
+		end
+	end
+
+	clear_holes(s_cookies, n)
+end
+
+local function get_expired(jar, time)
+	time = time or os.time()
+	local cookies = jar.cookies
+	local returned_cookies = {}
+	for i=#cookies, 1, -1 do
+		local cookie = cookies[i]
+		if cookie.expires > time then
+			break
+		end
+		returned_cookies[#returned_cookies + 1] = cookie
+	end
+	return returned_cookies
+end
+
+function cookiejar_methods:remove_expired(time)
+	self:remove_cookies(get_expired(self, time))
+end
+
+function cookiejar_methods:trim(size)
+	self:remove_expired()
+	local cookies = self.cookies
+	if #cookies > size then
+		for i=#cookies, size + 1, -1 do
+			local cookie = cookies[i]
+			cookies[i] = nil
+			local by_domain = cookies[cookie.domain]
+			local by_path = by_domain[cookie.path]
+			by_path[cookie.key] = nil
+			if not next(by_path) then
+				by_domain[cookie.path] = nil
+				if not next(by_domain) then
+					cookies[cookie.domain] = nil
+				end
+			end
+		end
+	end
+end
+
+local function serialize_cookies(cookies)
+	local out_values = {}
+	for _, cookie in pairs(cookies) do
+		out_values[#out_values + 1] = cookie.key .. "=" .. cookie.value
+	end
+	return table.concat(out_values, "; ")
+end
+
+function cookiejar_methods:serialize_cookies_for(domain, path, secure)
+	-- explicitly check for secure; the other two will fail if given bad args
+	assert(type(secure) == "boolean", "expected boolean for `secure`")
+
+	-- clear out expired cookies
+	self:remove_expired()
+
+	-- return empty table if no cookies are found
+	if not self.cookies[domain] then
+		return {}
+	end
+
+	-- check all paths and flatten into a list of sets
+	local sets = {}
+	for stored, set in pairs(self.cookies[domain]) do
+		if stored:sub(1, #path) == path then
+			for _, cookie in pairs(set) do
+				sets[#sets + 1] = cookie
+			end
+		end
+	end
+
+	-- sort as per RFC 6265 section 5.4 part 2; while it's not needed, it will
+	-- help with tests where values need to be reproducible
+	table.sort(sets, function(x, y)
+		if #x.path == #y.path then
+			return x.creation < y.creation
+		else
+			return #x.path > #y.path
+		end
+	end)
+
+	-- populate cookie list
+	local cookies = {}
+	for _, cookie in pairs(sets) do
+		if not cookie.host_only then
+			if self.psl_object:is_cookie_domain_acceptable(domain, cookie.domain) then
+				cookies[#cookies + 1] = cookie
+			end
+		elseif cookie.domain == domain then
+			cookies[#cookies + 1] = cookie
+		end
+	end
+
+	local n = #cookies
+	-- remove cookies requiring secure connections on insecure connections
+	for index, cookie in pairs(cookies) do
+		if cookie.secure and not secure then
```

Comment:
Do this check when adding; don't go through making holes that you later need to fixup...